docs/testplan/DWRR-Snappi-Test-Plan.md

```diff
@@ -0,0 +1,797 @@
+# DWRR Test Plan for Network Switches
+
+## 1. Introduction
+
+### 1.1 Purpose
+
+This test plan aims to validate the Deficit Weighted Round Robin (DWRR) scheduling functionality on network switches. DWRR is a scheduling algorithm that ensures fair allocation of bandwidth among different queues during periods of congestion, based on assigned weights to each queue.
+
+### 1.2 Test Environment Overview
+
+```mermaid
+flowchart TD
+    TA[Traffic Analyzer] -->|Physical Port 1<br>800G| SW(Switch Under Test)
+    TA -->|Physical Port 2<br>800G| SW
+    TA -->|Physical Port 3<br>800G| SW
+    TA -->|Physical Port 4<br>800G| SW
+    
+    subgraph "Traffic Analyzer"
+        P1L1[Logical Port 1.1 100G]
+        P1L2[Logical Port 1.2 100G]
+        P1L8[Logical Port 1.8 100G]
+        
+        P2L1[Logical Port 2.1 100G]
+        P2L2[Logical Port 2.2 100G]
+        P2L8[Logical Port 2.8 100G]
+        
+        P3L1[Logical Port 3.1100G]
+        P3L2[Logical Port 3.2100G]
+        P3L8[Logical Port 3.8100G]
+        
+        P4L1[Logical Port 4.1100G]
+        P4L2[Logical Port 4.2100G]
+        P4L8[Logical Port 4.8100G]
+    end
+    
+    style P1L8 fill:#fff,stroke-dasharray: 5 5
+    style P2L8 fill:#fff,stroke-dasharray: 5 5
+    style P3L8 fill:#fff,stroke-dasharray: 5 5
+    style P4L8 fill:#fff,stroke-dasharray: 5 5
+```
+
+*For all intents and purposes, the `Traffic Analyzer` (in the diagrams) is the ingress traffic generator port.*
+
+### 1.3 Equipment Requirements
+
+- **Device Under Test (DUT)**: 
+  - Network switch with DWRR scheduling capabilities
+- **Traffic Generator**: 
+  - Test equipment with snappi API support (Ixia, Keysight, etc.)
+  - 4 physical ports (800G each) with 8 logical ports (100G each) per physical port
+
+### 1.4 Pre-Test Requirements
+
+1. The switch must be configured with DWRR scheduling policies
+2. DSCP to queue mappings must be appropriately configured
+3. Traffic generator must be properly connected to the switch
+4. All interfaces must be in the "up" state
+
+### 1.5 General Test Methodology
+
+For each test scenario:
+
+1. Query the switch for existing DSCP-to-queue mappings and scheduler configurations
+2. Configure traffic flows with appropriate DSCP values based on switch configuration
+3. Generate traffic that causes congestion at the egress port
+4. Capture traffic at the egress port
+5. Analyze the distribution of traffic based on DSCP values
+6. Validate that the observed distribution matches the expected distribution based on DWRR weights (within 2% tolerance)
+
+## 2. Test Scenarios
+
+### 2.1 Scenario 1: Basic DWRR Validation with 2 Queues
+
+#### 2.1.1 Test Objective
+Validate DWRR functionality with 2 queues using standard weight ratios (60:40).
+
+#### 2.1.2 Topology
+
+```mermaid
+flowchart LR
+    subgraph "Traffic Generator"
+        TX1[TX Port 1 100GDSCP 160G]
```

Comment:
use % will be better than speed, because it makes the plan more generic and can work with other cases, such as 400G cases or even 800G cases.